File: Blind150/8.Heap_Priority Queue/Python/task_scheduler.py
# https://leetcode.com/problems/task-scheduler/description/
import heapq

from typing import List
from collections import Counter

class Solution:
    def leastInterval(self, tasks: List[str], n: int) -> int:
        count = Counter(tasks)

        next_task_map = {}

        t = 0

        total_task = len(count.keys())

        maxHeap = [-cnt for cnt in count.values()]

        heapq.heapify(maxHeap)

        while total_task !=0:

            t += 1

            if nt := next_task_map.get(t):
                heapq.heappush(maxHeap, nt)

            if maxHeap:
                ct = heapq.heappop(maxHeap)

                if ct + 1 == 0:
                    total_task -= 1
                else:
                    next_task_map[t+n+1] = ct + 1


        return t

File: Blind150/8.Heap_Priority Queue/Go/task_scheduler.go
// https://leetcode.com/problems/task-scheduler/description/
type MaxHeap []int

func (h MaxHeap) Len() int           { return len(h) }
func (h MaxHeap) Less(i, j int) bool { return h[i] > h[j] }
func (h MaxHeap) Swap(i, j int)      { h[i], h[j] = h[j], h[i] }

func (h *MaxHeap) Push(x any) {

	*h = append(*h, x.(int))
}

func (h *MaxHeap) Pop() any {
	old := *h
	n := len(old)
	x := old[n-1]
	*h = old[:n-1]

	return x

}

func leastInterval(tasks []byte, n int) int {

	// Step 1: Count task frequencies
	taskArrival := map[string]int{}
	for _, task := range tasks {
		taskArrival[string(task)]++
	}

	// Step 2: Initialize other variables
	nextTaskMap := map[int]int{}
	t := 0
	totalTaskCount := len(taskArrival)

	// Step 3: Initialize the MaxHeap
	maxHeap := &MaxHeap{}
	heap.Init(maxHeap)

	for _, count := range taskArrival {
		heap.Push(maxHeap, count)
	}

	for totalTaskCount > 0 {

		t += 1

		if taskValue, exists := nextTaskMap[t]; exists {
			heap.Push(maxHeap, taskValue)
		}

		if maxHeap.Len() > 0 {

			ct := heap.Pop(maxHeap).(int)

			if ct-1 == 0 {
				totalTaskCount -= 1
			} else {
				nextTaskMap[t+n+1] = ct - 1
			}
		}
	}

	return t

}
